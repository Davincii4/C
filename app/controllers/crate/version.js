import { alias, readOnly, gt } from '@ember/object/computed';
import { inject as service } from '@ember/service';
import Controller from '@ember/controller';
import PromiseProxyMixin from '@ember/object/promise-proxy-mixin';
import ArrayProxy from '@ember/array/proxy';
import { computed, observer } from '@ember/object';
import { later } from '@ember/runloop';
import $ from 'jquery';
import moment from 'moment';

const NUM_VERSIONS = 5;

const PromiseArray = ArrayProxy.extend(PromiseProxyMixin);

export default Controller.extend({
    session: service(),

    isDownloading: false,

    downloadsContext: computed('requestedVersion', 'model', 'crate', function() {
        return this.requestedVersion ? this.model : this.crate;
    }),
    downloads: alias('downloadsContext.version_downloads'),
    extraDownloads: alias('downloads.content.meta.extra_downloads'),

    fetchingFollowing: true,
    following: false,
    currentVersion: alias('model'),
    requestedVersion: null,
    keywords: alias('crate.keywords'),
    categories: alias('crate.categories'),
    badges: alias('crate.badges'),
    isOwner: computed('crate.owner_user', 'session.currentUser.login', function() {
        return this.get('crate.owner_user').findBy('login', this.get('session.currentUser.login'));
    }),

    sortedVersions: readOnly('crate.versions'),

    smallSortedVersions: computed('sortedVersions', function() {
        return this.sortedVersions.slice(0, NUM_VERSIONS);
    }),

    hasMoreVersions: gt('sortedVersions.length', NUM_VERSIONS),

    displayedAuthors: computed('currentVersion.authors.[]', function() {
        return PromiseArray.create({
            promise: this.get('currentVersion.authors').then(authors => {
                let ret = authors.slice();
                let others = authors.get('meta');
                for (let i = 0; i < others.names.length; i++) {
                    ret.push({ name: others.names[i] });
                }
                return ret;
            }),
        });
    }),

    anyKeywords: gt('keywords.length', 0),
    anyCategories: gt('categories.length', 0),

    currentDependencies: computed('currentVersion.dependencies', function() {
        let deps = this.get('currentVersion.dependencies');

        if (deps === null) {
            return [];
        }

        return PromiseArray.create({
            promise: deps.then(deps => {
                return deps.filter(dep => dep.get('kind') !== 'dev').uniqBy('crate_id');
            }),
        });
    }),

    currentDevDependencies: computed('currentVersion.dependencies', function() {
        let deps = this.get('currentVersion.dependencies');
        if (deps === null) {
            return [];
        }
        return PromiseArray.create({
            promise: deps.then(deps => {
                return deps.filterBy('kind', 'dev');
            }),
        });
    }),

    downloadData: computed('downloads', 'extraDownloads', 'requestedVersion', function() {
        let downloads = this.downloads;
        if (!downloads) {
            return;
        }

        let extra = this.extraDownloads || [];

        let dates = {};
        let versions = [];
        for (let i = 0; i < 90; i++) {
            let now = moment().subtract(i, 'days');
            dates[now.format('MMM D')] = { date: now, cnt: {} };
        }

        downloads.forEach(d => {
            let version_id = d.get('version.id');
            let key = moment(d.get('date'))
                .utc()
                .format('MMM D');
            if (dates[key]) {
                let prev = dates[key].cnt[version_id] || 0;
                dates[key].cnt[version_id] = prev + d.get('downloads');
            }
        });

        extra.forEach(d => {
            let key = moment(d.date)
                .utc()
                .format('MMM D');
            if (dates[key]) {
                let prev = dates[key].cnt[null] || 0;
                dates[key].cnt[null] = prev + d.downloads;
            }
        });
        if (this.requestedVersion) {
            versions.push(this.model.getProperties('id', 'num'));
        } else {
            this.smallSortedVersions.forEach(version => {
                versions.push(version.getProperties('id', 'num'));
            });
        }
        if (extra.length > 0) {
            versions.push({
                id: null,
                num: 'Other',
            });
        }

        let headers = ['Date'];
        versions.sort(b => b.num).reverse();
        for (let i = 0; i < versions.length; i++) {
            headers.push(versions[i].num);
        }
        let data = [headers];
        for (let date in dates) {
            let row = [dates[date].date.toDate()];
            for (let i = 0; i < versions.length; i++) {
                row.push(dates[date].cnt[versions[i].id] || 0);
            }
            data.push(row);
        }

        return data;
    }),

    toggleClipboardProps(isSuccess) {
        this.setProperties({
            showSuccess: isSuccess,
            showNotification: true,
        });
        later(
            this,
            () => {
                this.set('showNotification', false);
            },
            2000,
        );
    },

    actions: {
        copySuccess(event) {
            event.clearSelection();
            this.toggleClipboardProps(true);
        },

        copyError() {
            this.toggleClipboardProps(false);
        },

        toggleFollow() {
            this.set('fetchingFollowing', true);

            let crate = this.crate;
            let op = this.toggleProperty('following') ? crate.follow() : crate.unfollow();

            return op.finally(() => this.set('fetchingFollowing', false));
        },

        showDeps() {
            var margin = {top: window.innerHeight / 100 * 3,
                          right: (window.innerWidth / 100 * 6),
                          bottom: window.innerHeight / 100 * 3,
                          left: (window.innerWidth / 100 * 6)},
                width = 1950,
                height = window.innerHeight - (window.innerHeight / 100 * 6);

            var i = 0,
                duration = 750,
                root = {
                    crate_id: this.get('crate.name'),
                    version_id: this.get('currentVersion.num'),
                    level: 1,
                };

            var tree = d3.layout.tree().size([height, width]);

            var diagonal = d3.svg.diagonal()
                                 .projection(function(d) {
                                     return [d.y, d.x];
                                 });

            var svg = d3.select("#deps-rendering")
                        .attr("width", width)
                        .attr("height", height)
                        .append("g")
                        .attr("transform", "translate(" + margin.left + "," + margin.top + ")");

            // Load dependencies.
            //root = flare;
            root.x0 = height / 2;
            root.y0 = 0;

            var total = 0;
            var crates = {};

            function get_wanted_version_num(object, wanted_version) {
                for (var i = 0; i < object.versions.length; ++i) {
                    if (object.versions[i].id == wanted_version) {
                        return object.versions[i].num;
                    }
                }
                return object.versions[0].num;
            }

            function decrement_total(level) {
                total -= 1;
                /*if (Object.keys(crates).length % 20 === 0) {
                    console.log(crates);
                }*/
                if (total === 0 || level < 3) {
                    update(root);
                    if (total === 0) {
                        console.log(root);
                    }
                }
            }

            function add_dependencies(e, children) {
                for (var i = 0; i < children.length; ++i) {
                    children[i].level = e.level + 1;
                }
                e.children = null;
                e._children = children;
                if (e._children && e.level < 4) {
                    e._children.forEach(collapse);
                }
                if (e.level < 3) {
                    e.children = e._children;
                    e._children = null;
                }
                decrement_total(e.level);
            }

            var all_urls = {};

            function get_next_url(url, callback) {
                if (all_urls[url] === undefined) {
                    all_urls[url] = {'callbacks': [callback], 'pending': true};
                    d3.json(url, (error, flare) => {
                        if (error) {
                            decrement_total();
                            all_urls[url] = undefined;
                            return;
                        }
                        var callbacks = all_urls[url].callbacks;
                        all_urls[url] = flare;
                        all_urls[url].pending = false;
                        for (var i = 0; i < callbacks.length; ++i) {
                            callbacks[i](flare);
                        }
                    });
                } else if (all_urls[url].pending === true) {
                    all_urls[url].callbacks.push(callback);
                } else {
                    callback(all_urls[url]);
                }
            }

            function get_dependencies_for_version(crate, version, e) {
                var tmp = e;
                get_next_url(`/api/v1/crates/${crate}/${version}/dependencies`, (data) => {
                    tmp.version_id = version;
                    add_dependencies(tmp, JSON.parse(JSON.stringify(data.dependencies)));
                });
            }

            function collapse(d) {
                if (!d.children && d.optional !== true) {
                    total += 1;
                    if (d.version_id.toString().indexOf(".") !== -1) {
                        get_dependencies_for_version(d.crate_id, d.version_id, d);
                    } else {
                        get_next_url(`/api/v1/crates/${d.crate_id}`, (data) => {
                            var ver = get_wanted_version_num(data, d.version_id);
                            get_dependencies_for_version(d.crate_id, ver, d);
                        });
                    }
                }
            }

            collapse(root);

            //d3.select(self.frameElement).style("height", "800px");

            function update(source) {
                // Compute the new tree layout.
                var nodes = tree.nodes(source).reverse(),
                    links = tree.links(nodes);

                // Normalize for fixed-depth.
                nodes.forEach(function(d) {
                    d.y = d.depth * 180;
                });

                // Update the nodes...
                var node = svg.selectAll("g.node")
                              .data(nodes, function(d) {
                                  return d.id || (d.id = ++i);
                              });

                // Enter any new nodes at the parent's previous position.
                var nodeEnter = node.enter()
                                    .append("g")
                                    .attr("class", "node")
                                    .attr("transform", function(d) {
                                        return "translate(" + source.y0 + "," + source.x0 + ")";
                                    })
                                    .on("click", click);

                nodeEnter.append("circle")
                         .attr("r", 1e-6)
                         .style("fill", function(d) {
                             return d._children ? "lightsteelblue" : "#fff";
                         });

                nodeEnter.append("text")
                         .attr("x", function(d) {
                             return d.children || d._children ? -10 : 10;
                         })
                         .attr("dy", ".35em")
                         .attr("text-anchor", function(d) {
                             return d.children || d._children ? "end" : "start";
                         })
                         .text(function(d) {
                             return d.crate_id + ' ' + d.version_id;
                         })
                         .style("fill-opacity", 1e-6);

                // Transition nodes to their new position.
                var nodeUpdate = node.transition()
                                     .duration(duration)
                                     .attr("transform", function(d) {
                                         return "translate(" + d.y + "," + d.x + ")";
                                     });

                nodeUpdate.select("circle")
                          .attr("r", 4.5)
                          .style("fill", function(d) {
                              return d._children ? "lightsteelblue" : "#fff";
                          });

                nodeUpdate.select("text")
                          .style("fill-opacity", 1);

                // Transition exiting nodes to the parent's new position.
                var nodeExit = node.exit()
                                   .transition()
                                   .duration(duration)
                                   .attr("transform", function(d) {
                                       return "translate(" + source.y + "," + source.x + ")";
                                   })
                                   .remove();

                nodeExit.select("circle")
                        .attr("r", 1e-6);

                nodeExit.select("text")
                        .style("fill-opacity", 1e-6);

                // Update the links…
                var link = svg.selectAll("path.link")
                              .data(links, function(d) { return d.target.id; });

                // Enter any new links at the parent's previous position.
                link.enter()
                    .insert("path", "g")
                    .attr("class", "link")
                    .attr("d", function(d) {
                        var o = {x: source.x0, y: source.y0};
                        return diagonal({source: o, target: o});
                    });

                // Transition links to their new position.
                link.transition()
                    .duration(duration)
                    .attr("d", diagonal);

                // Transition exiting nodes to the parent's new position.
                link.exit()
                    .transition()
                    .duration(duration)
                    .attr("d", function(d) {
                        var o = {x: source.x, y: source.y};
                        return diagonal({source: o, target: o});
                    })
                    .remove();

                // Stash the old positions for transition.
                nodes.forEach(function(d) {
                    d.x0 = d.x;
                    d.y0 = d.y;
                });
            }

            // Toggle children on click.
            function click(d) {
                if (d.children) {
                    d._children = d.children;
                    d.children = null;
                } else {
                    d.children = d._children;
                    d._children = null;
                }
                update(root);
            }
            this.set('showingDeps', true);
        }
    },

    report: observer('crate.readme', function() {
        setTimeout(() => $(window).trigger('hashchange'));
    }),
});
