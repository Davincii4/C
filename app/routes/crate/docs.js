import Route from '@ember/routing/route';
import { inject as service } from '@ember/service';

export default Route.extend({
    flashMessages: service(),

    redirect() {
        const crate = this.modelFor('crate');

        const documentation = crate.get('documentation');
        if (documentation) {
            window.location = documentation;
        } else {
            // Redirect to the crate's main page and show a flash error if
            // no documentation is found
            const message = 'Crate does not supply a documentation URL';
            this.flashMessages.queue(message);
            this.replaceWith('crate', crate);
        }
    },
});
