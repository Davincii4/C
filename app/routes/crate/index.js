import Ember from 'ember';

export default Ember.Route.extend({
    beforeModel() {
        const url = this.get('router.url');
        if (url.startsWith('/crates/') && !url.endsWith('/')) {
            this.replaceWith('crate.version', '');
        } else {
            this.transitionTo('crate.version', '');
        }
    }
});
