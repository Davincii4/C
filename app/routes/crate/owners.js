import Route from '@ember/routing/route';

export default Route.extend({
    setupController(controller) {
        this._super(...arguments);
        const crate = this.modelFor('crate');
        controller.set('crate', crate);
    },
});
