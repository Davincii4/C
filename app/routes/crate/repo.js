import Route from '@ember/routing/route';
import { inject as service } from '@ember/service';

export default class CrateRepoRoute extends Route {
  @service flashMessages;
  @service redirector;

  redirect() {
    let crate = this.modelFor('crate');

    let repository = crate.get('repository');
    if (repository) {
      this.redirector.redirectTo(repository);
    } else {
      // Redirect to the crate's main page and show a flash error if
      // no repository is found
      let message = 'Crate does not supply a repository URL';
      this.flashMessages.queue(message);
      this.replaceWith('crate', crate);
    }
  }
}
