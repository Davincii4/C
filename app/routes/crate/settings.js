import Route from '@ember/routing/route';

export default class SettingsRoute extends Route {
  async model() {
    let crate = this.modelFor('crate');
    let [users, teams] = await Promise.all([crate.owner_user, crate.owner_team]);
    return { crate, teams, users };
  }
}
