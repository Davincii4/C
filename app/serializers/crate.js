import DS from 'ember-data';

export default DS.RESTSerializer.extend({
    isNewSerializerAPI: true,
    attrs: {
        originalDocumentation: 'documentation'
    },

    extractRelationships(modelClass, resourceHash) {
        if (resourceHash.versions == null) {
            delete resourceHash.versions;
        }

        return this._super(...arguments);
    }
});
