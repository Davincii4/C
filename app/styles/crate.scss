#crates-heading {
    width: 100%;
    padding: 20px;
    background-color: darken($main-bg, 6%);

    .crates-heading-name {
        img, h1, h2, .name {display: inline-block; vertical-align: middle; }
    }
    .info {
        @include align-items(center);
    }
    h1 { padding-left: 10px; }
    h2 { color: $blurb-fg; padding-left: 10px; padding-top: 10px; font-size: 1.5em; }
    .right {
        @include flex(2);
        @include display-flex;
        @include flex-direction(row-reverse);
        @include align-items(center);
    }
    @media only screen and (max-width: 650px) {
        .right { display: none; }
    }

    &.crate-index, &.crate-search {
        @include flex-wrap(wrap);
        .magnifier {
            width: 32px;
            height: 32px;
        }
    }
}

#selection {
    width: 100%;
    @include display-flex;
    @include justify-content(space-between);
    padding: 20px;
    @include border-radius(5px);
    background-color: white;
    margin-bottom: 40px;

    a {
        text-decoration: none;
        font-weight: bold;
        color: inherit;
        padding: 5px;
    }
    a:hover { background-color: $main-bg-dark; }
    a.active { background-color: $main-bg-dark; }

    select {
        display: none;
        width: 100%;
        padding: 5px;
    }

    @media only screen and (max-width: 650px) {
        padding: 0;
        background-color: inherit;
        a { display: none; }
        select { display: block; }
    }
}

#results {
    width: 100%;
    @include display-flex;
    @include justify-content(space-between);
    margin-bottom: 25px;

    .cur, .total { color: $main-color; font-weight: bold; }

    .dropdown-container { font-size: 85%; }
    a.dropdown {
        background-color: $main-bg-dark;
        padding: 10px;
        display: inline-block;
        @include border-radius(5px);
    }
}

.white-rows {
    background-color: white;
    padding: 0 20px;
    margin-bottom: 20px;
    @include display-flex;
    @include flex-direction(column);

    .row {
        width: 100%;
        border-bottom: 2px solid $gray-border;
        padding: 20px 0;
        @include display-flex;
        @include justify-content(space-between);

        &:last-of-type { border: none; }
    }
}

.crate {
    .desc {
        padding-top: 5px;
        @include display-flex;
        @include flex-direction(column);
        width: 85%;
    }

    .info a {
        color: $main-color;
        font-weight: bold;
        text-decoration: none;
        font-size: 120%;
    }
    .vers { margin-left: 10px; }

    .stats {
        width: 15%;
        color: $main-color-light;
    }
    .downloads { @include display-flex; @include align-items(center); }
}

.pagination {
    text-align: center;
    font-size: 90%;
    margin-bottom: 20px;

    a {
        color: $main-color-light;
        text-decoration: none;
        padding: 5px 6px;
    }
    a:hover { background-color: $main-bg-dark; }
    a.active { background-color: $main-bg-dark; }
}

.about {
    font-size: 1.3em;
    vertical-align: middle;
    text-align: justify;
    p { margin: auto auto; }
}

.install {
    font-size: 120%;
    margin: 0px;
    padding: 10px;
    background-color: darken($main-bg, 3%);
    @include display-flex;

    .action {
        display: none;
        text-align: center;
        color: $header-bg;
        font-family: times;
        font-size: 1.4em;
    }
    code {
        background-color: $main-bg;
        border-radius: 10px;
        color: black;
        font-size: 0.9em;
        padding: 10px;
        padding-left: 30px;
    }

    @media only screen and (min-width: 500px) {
        .action { @include flex(2); display: block; }
        code { @include flex(8); }
    }
}

.crate-info {
    padding-bottom: 20px;
    border-bottom: 5px solid $gray-border;
    margin-bottom: 20px;

    a {
        color: $header-bg;
    }

    h3 {
        font-weight: bold;
    }

    @media only screen and (min-width: 820px) {
        @include flex-direction(row);

        .docs {
            vertical-align: middle;
            text-align: center;
            padding: 10px;
        }
        .authorship {
            @include flex(3);

            .top, .bottom {
                @include flex-direction(column);
            }

        }
    }

    @media only screen and (max-width: 480px) {
        .authorship {
            .top, .bottom {
                @include flex-direction(column);
            }
        }
    }

    .authorship {
        .top, .bottom {
            @include display-flex;

            > * { @include flex(1); }
        }

    }

    ul { list-style: none; padding: 0px; }

    ul.owners, ul.keywords {
        @include display-flex;
        @include flex-wrap(wrap);
        list-style: none;
        padding: 0;
        margin: 0;
        li { margin: 0 7px 7px 0; }
        img { border-radius: 5px; width: 50px; height: 50px; }
    }


    .last-update {
        color: $main-color-light;
        font-size: 90%;
        line-height: 25px;
    }
    .date {
        font-weight: bold;
        margin-bottom: 40px;
    }
}

#crate-downloads {
    @include display-flex;
    @include flex-wrap(wrap);

    h3 {
        width: 100%;
        font-weight: bold;
        font-size: 2em;
        color: $header-bg;
        margin: 20px 0px 30px 30px;
    }

    .stat {
        border-left: 1px solid $gray-border;
        padding: 10px 20px;

        .num {
            font-size: 160%;
            font-weight: bold;
            margin-bottom: 4px;
            color: $blurb-fg;
        }
    }

    .stats {
        @include flex-grow(7);
        @include display-flex;
        @include flex-wrap(wrap);
    }
    .versions {
        margin-top: 50px;
        @include flex-grow(3);
        text-align: center;
        @include display-flex;
        @include flex-direction(column);
        .small { padding: 10px; }
    }

    .button-holder {
        position: relative;
    }

    .graph {
        @include flex-grow(10);
        width: 100%;
        margin: 30px 0;
        padding-bottom: 20px;
        background-color: $main-bg;

        h4 { color: $main-color-light; }
    }
    .graph-data {
        width: 100%;
        height: 500px;
    }
}

#crate-links {
    @include display-flex;
    padding-bottom: 20px;
    border-bottom: 5px solid $gray-border;
    margin-bottom: 30px;

    h3 { margin-bottom: 5px; }
    .section {
        @include flex(1);
        li { line-height: 24px; }
        .date { color: $main-color-light; padding-left: 5px; }
    }

    ul { list-style: none; padding: 0; }

    @media only screen and (max-width: 380px) {
        @include flex-direction(column);
    }
    @media only screen and (max-width: 730px) {
        #crate-dev-dependencies { display: none; }
    }
    @media only screen and (max-width: 430px) {
        #crate-dependencies { display: none; }
    }
}

.all-versions-back { margin-bottom: 20px; }

#crate-all-versions {
    margin-top: 10px;

    .small { margin-left: 20px; display: inline-block; }
    a.arrow {
        display: inline-block;
        float: right;
    }
}
