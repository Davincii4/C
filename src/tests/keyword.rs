use crate::insta::rfc3339_redaction;
use crate::{builders::CrateBuilder, RequestHelper, TestApp};
use cargo_registry::{models::Keyword, views::EncodableKeyword};
use insta::assert_json_snapshot;
use serde_json::Value;

#[derive(Deserialize)]
struct GoodKeyword {
    keyword: EncodableKeyword,
}

#[test]
fn index() {
    let url = "/api/v1/keywords";
    let (app, anon) = TestApp::init().empty();
    let json: Value = anon.get(url).good();
    assert_json_snapshot!(json);

    app.db(|conn| {
        Keyword::find_or_create_all(conn, &["foo"]).unwrap();
    });

    let json: Value = anon.get(url).good();
    assert_json_snapshot!(json, { ".**.created_at" => rfc3339_redaction() });
}

#[test]
fn show() {
    let url = "/api/v1/keywords/foo";
    let (app, anon) = TestApp::init().empty();
    anon.get(url).assert_not_found();

    app.db(|conn| {
        Keyword::find_or_create_all(conn, &["foo"]).unwrap();
    });

    let json: Value = anon.get(url).good();
    assert_json_snapshot!(json, { ".**.created_at" => rfc3339_redaction() });
}

#[test]
fn uppercase() {
    let url = "/api/v1/keywords/UPPER";
    let (app, anon) = TestApp::init().empty();
    anon.get(url).assert_not_found();

    app.db(|conn| {
        Keyword::find_or_create_all(conn, &["UPPER"]).unwrap();
    });

    let json: Value = anon.get(url).good();
    assert_json_snapshot!(json, { ".**.created_at" => rfc3339_redaction() });
}

#[test]
fn update_crate() {
    let (app, anon, user) = TestApp::init().with_user();
    let user = user.as_model();

    let cnt = |kw: &str| {
        let json: GoodKeyword = anon.get(&format!("/api/v1/keywords/{}", kw)).good();
        json.keyword.crates_cnt as usize
    };

    let krate = app.db(|conn| {
        Keyword::find_or_create_all(conn, &["kw1", "kw2"]).unwrap();
        CrateBuilder::new("fookey", user.id).expect_build(conn)
    });

    app.db(|conn| {
        Keyword::update_crate(conn, &krate, &[]).unwrap();
    });
    assert_eq!(cnt("kw1"), 0);
    assert_eq!(cnt("kw2"), 0);

    app.db(|conn| {
        Keyword::update_crate(conn, &krate, &["kw1"]).unwrap();
    });
    assert_eq!(cnt("kw1"), 1);
    assert_eq!(cnt("kw2"), 0);

    app.db(|conn| {
        Keyword::update_crate(conn, &krate, &["kw2"]).unwrap();
    });
    assert_eq!(cnt("kw1"), 0);
    assert_eq!(cnt("kw2"), 1);

    app.db(|conn| {
        Keyword::update_crate(conn, &krate, &[]).unwrap();
    });
    assert_eq!(cnt("kw1"), 0);
    assert_eq!(cnt("kw2"), 0);

    app.db(|conn| {
        Keyword::update_crate(conn, &krate, &["kw1", "kw2"]).unwrap();
    });
    assert_eq!(cnt("kw1"), 1);
    assert_eq!(cnt("kw2"), 1);

    app.db(|conn| {
        Keyword::update_crate(conn, &krate, &[]).unwrap();
    });
    assert_eq!(cnt("kw1"), 0);
    assert_eq!(cnt("kw2"), 0);
}
