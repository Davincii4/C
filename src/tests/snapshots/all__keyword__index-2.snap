---
source: src/tests/keyword.rs
expression: json
---
{
  "keywords": [
    {
      "crates_cnt": 0,
      "created_at": "[datetime]",
      "id": "foo",
      "keyword": "foo"
    }
  ],
  "meta": {
    "total": 1
  }
}
